core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/TypeUtils.kt

```diff
@@ -390,14 +392,24 @@ internal fun <T> getValuesType(values: List<T>, type: KType, infer: Infer): KTyp
  * @param listifyValues if true, then values and nulls will be wrapped in a list if they appear among other lists.
  *   For example: `[1, null, listOf(1, 2, 3)]` will become `List<Int>` instead of `Any?`
  *   Note: this parameter is ignored if another [Collection] is present in the values.
+ * @param allColsMakesRow if true, then, if all values are non-null same-sized columns, we assume
+ *   that a column group should be created instead of a [DataColumn][DataColumn]`<`[AnyCol][AnyCol]`>`,
+ *   so the function will return [DataRow].
  */
 @PublishedApi
-internal fun guessValueType(values: Sequence<Any?>, upperBound: KType? = null, listifyValues: Boolean = false): KType {
+internal fun guessValueType(
+    values: Sequence<Any?>,
+    upperBound: KType? = null,
+    listifyValues: Boolean = false,
+    allColsMakesRow: Boolean = false,
```

Comment:
In all other cases, allColsMakesRow can be false. I put it here so that we don't check all values for types multiple times in different places, like we did before, we just have this one place which makes the behavior easy to reason about and debug. `listifyValues` is a similar case, it was only used in Json I believe? (though I'm not sure it still is). `guessValueType()` is already a very difficult function, however, all the logic of "values -> columnType" for the entire library lives here so let's not make it more difficult by spreading logic across multiple files and then end up with another `createColumn` function which works exactly like `createColumnGuessingType` except when it doesn't